### PR TITLE
mgmt: mcumgr: grp: img_mgmt: Add image data written callback

### PR DESCRIPTION
**Pull request by nordicjm**
Adds an optional callback upon image data being written, can be
    used for syncing or timeout purposes.

and

    Adds a function which gets the callback ID index of a given
    event ID.

### CODE REVIEW

**Review by de-nordic — DISMISSED** [review]

File: subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt.c [hunk 1]
```diff
@@ -642,6 +642,10 @@ img_mgmt_upload(struct smp_streamer *ctxt)
 #if defined(CONFIG_MCUMGR_GRP_IMG_STATUS_HOOKS)
 			(void)mgmt_callback_notify(MGMT_EVT_OP_IMG_MGMT_DFU_PENDING, NULL, 0,
 						   &ret_rc, &ret_group);
+		} else {
+			/* Notify that the write has completed */
+			(void)mgmt_callback_notify(MGMT_EVT_OP_IMG_MGMT_DFU_WRITE_COMPLETE, NULL,
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt.c:647]: For the same reason, I would suggest to call it also if an error happens after application was notified that data is going to be stored (e.g. on `img_mgmt_erase_image_data` error) and if SHA matches so write would not be performed

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: subsys/mgmt/mcumgr/mgmt/src/mgmt.c [hunk 2]
```diff
@@ -143,6 +144,28 @@ enum mgmt_cb_return mgmt_callback_notify(uint32_t event, void *data, size_t data
 
 	return return_status;
 }
+
+uint16_t mgmt_evt_get_index(uint32_t event)
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply subsys/mgmt/mcumgr/mgmt/src/mgmt.c]: Fixed

File: include/zephyr/mgmt/mcumgr/mgmt/callbacks.h [hunk 3]
```diff
@@ -165,6 +165,9 @@ enum img_mgmt_group_events {
 	/** Callback when an image has been confirmed. */
 	MGMT_EVT_OP_IMG_MGMT_DFU_CONFIRMED	= MGMT_DEF_EVT_OP_ID(MGMT_EVT_GRP_IMG, 4),
 
+	/** Callback when an image write command has finished writing to flash. */
+	MGMT_EVT_OP_IMG_MGMT_DFU_WRITE_COMPLETE	= MGMT_DEF_EVT_OP_ID(MGMT_EVT_GRP_IMG, 5),
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply include/zephyr/mgmt/mcumgr/mgmt/callbacks.h]: Renamed

**Review by de-nordic — APPROVED** [review]

File: subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt.c]: Discussed offline. We can leave it as is.

**Review by MarekPieta — APPROVED** [review]

**Comment by nordicjm** [comment]
@fabiobaltieri could you please merge?

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**